### PR TITLE
Proxy with keepalive support

### PR DESCRIPTION
**Pull request by pxjohnny**

### CODE REVIEW

File: lib/px/utils/pxclient.lua [hunk 1]
```diff
@@ -186,7 +186,7 @@ function M.load(px_config)
     -- @allow_failure - will allow http status >= 400
     --
     -- @return - boolean value, success or failure
-    local function forward_to_perimeterx(server, port_overide, allow_failure)
+    local function forward_to_perimeterx(server, port_overide, allow_failure, pool_key)
```
Comment (barakpx) [inline lib/px/utils/pxclient.lua]: missing code doc for pool_key

File: lib/px/utils/pxclient.lua [hunk 2]
```diff
@@ -261,7 +261,7 @@ function M.load(px_config)
         ngx_req_set_uri(px_request_uri)
         px_common_utils.clear_first_party_sensitive_headers(px_config.sensitive_headers)
 
-        forward_to_perimeterx(px_config.client_host, px_config.client_port_overide, true)
+        forward_to_perimeterx(px_config.client_host, px_config.client_port_overide, true, "px_fp_client")
```
Comment (barakpx) [inline lib/px/utils/pxclient.lua]: can you change to px_client - everything if first party in the enforcer

File: lib/px/utils/pxclient.lua [hunk 3]
```diff
@@ -284,7 +284,7 @@ function M.load(px_config)
         ngx_req_set_uri(px_request_uri)
 
         px_common_utils.clear_first_party_sensitive_headers(px_config.sensitive_headers)
-        forward_to_perimeterx(px_config.captcha_script_host, nil, true)
+        forward_to_perimeterx(px_config.captcha_script_host, nil, true, "px_fp_captcha")
```
Comment (barakpx) [inline lib/px/utils/pxclient.lua]: same here - px_captcha

File: lib/px/utils/pxclient.lua [hunk 4]
```diff
@@ -334,7 +334,7 @@ function M.load(px_config)
             xff_header = ngx.var.remote_addr
         end
         ngx_req_set_header('X-Forwarded-For', xff_header)
-        local status = forward_to_perimeterx(px_config.collector_host, px_config.collector_port_overide, false)
+        local status = forward_to_perimeterx(px_config.collector_host, px_config.collector_port_overide, false, "px_fp_xhr")
```
Comment (barakpx) [inline lib/px/utils/pxclient.lua]: px_xhr

File: lib/px/utils/pxcommonutils.lua [hunk 5]
```diff
@@ -153,14 +153,68 @@ function _M.extract_cookie_names(cookies)
     return t
 end
 
-function _M.call_px_server(httpc, host, port, proxy)
-    if proxy ~= nil then
-        return httpc:connect_proxy(proxy, "https", host, port, "")
+function _M.call_px_server(httpc, host, port, px_config, pool_key)
+    if px_config.proxy_url ~= nil then
+        local scheme = px_config.ssl_enabled and "https" or "http"
```
Comment (barakpx) [inline lib/px/utils/pxcommonutils.lua]: we should remove 'ssl_enabled'  - it is a mess when you set boolean value and need to maintain url and port that correlate that decision

File: lib/px/utils/pxcommonutils.lua [hunk 6]
```diff
@@ -153,14 +153,68 @@ function _M.extract_cookie_names(cookies)
...
+        local proxy_authorization = px_config.proxy_authorization == nil and "" or px_config.proxy_authorization
```
Comment (barakpx) [inline lib/px/utils/pxcommonutils.lua]: why do we need to do that here? I thought when we set the px configuration we set the default or you what the customer configured - we can set empty string there.

Comment (pxjohnny) [reply lib/px/utils/pxcommonutils.lua]: yea the default is nil...

File: lib/px/utils/pxcommonutils.lua (hunk 5 above)
Comment (pxjohnny) [reply lib/px/utils/pxcommonutils.lua]: we should remove this in the next major version for sure

**Pull request merged**